packages/studio-plugin/src/utils/ComponentTreeHelpers.ts

```diff
@@ -175,15 +175,53 @@ export default class ComponentTreeHelpers {
     ancestorState: ComponentState,
     componentTree: ComponentState[]
   ) {
-    const descendants = ComponentTreeHelpers.mapComponentTree<ComponentState[]>(
-      componentTree,
-      (componentState, mappedChildren) => [
-        componentState,
-        ...mappedChildren.flat(),
-      ],
-      ancestorState
-    ).flat();
+    const descendants =
+      ComponentTreeHelpers.mapComponentTreeParentsFirst<ComponentState>(
+        componentTree,
+        (componentState) => componentState,
+        ancestorState
+      );
 
     return descendants;
   }
+
+  /**
+   * Returns the component UUID that is lowest common parent of the two target components.
+   * This is similar to the lowest common ancestor except it cannot be one of the targets.
+   */
+  static getLowestParentUUID(
+    targetOneUUID: string,
+    targetTwoUUID: string,
+    componentTree: ComponentState[]
+  ) {
+    if (!targetOneUUID || !targetTwoUUID) return undefined;
+    const componentOne = componentTree.find((c) => c.uuid === targetOneUUID);
+    const componentTwo = componentTree.find((c) => c.uuid === targetTwoUUID);
+    return ComponentTreeHelpers.getLowestCommonAncestorComponentUUID(
+      componentOne?.parentUUID,
+      componentTwo?.parentUUID,
+      componentTree
+    );
+  }
+
+  private static getLowestCommonAncestorComponentUUID(
+    targetOneUUID: string | undefined,
+    targetTwoUUID: string | undefined,
+    componentTree: ComponentState[]
+  ): string | undefined {
+    let LCA_UUID;
+    ComponentTreeHelpers.mapComponentTree(componentTree, (c) => {
+      if (LCA_UUID) return;
+      if (!LCA_UUID && targetOneUUID === targetTwoUUID) {
```

Comment:
nit: we don't need to explicitly check for `!LCA_UUID` now that we return early if `LCA_UUID` is truthy

Reply:
Done